dali/operators/color/brightness_contrast.cc

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/color/brightness_contrast.h"
+#include "dali/kernels/imgproc/color_manipulation/brightness_contrast.h"
+
+namespace dali {
+namespace {
+
+template <typename Out, typename In>
+using TheKernel = kernels::BrightnessContrastCpu<Out, In, 3>;
+
+}  // namespace
+
+
+DALI_SCHEMA(BrightnessContrast)
+                .DocStr(R"code(Change the brightness and contrast of the image.
+Additionally, this operator can change the type of data.)code")
+                .NumInput(1)
+                .NumOutput(1)
+                .AddOptionalArg(brightness_contrast::kBrightness,
+                                R"code(Set additive brightness delta. 0 denotes no-op)code", .0f,
+                                true)
+                .AddOptionalArg(brightness_contrast::kContrast,
+                                R"code(Set multiplicative contrast delta. 1 denotes no-op)code",
+                                1.f, true)
+                .AddOptionalArg(brightness_contrast::kOutputType,
+                                R"code(Set output data type)code", DALI_INT16);
+
+DALI_REGISTER_OPERATOR(BrightnessContrast, BrightnessContrastCpu, CPU)
+
+
+bool BrightnessContrastCpu::SetupImpl(std::vector<OutputDesc> &output_desc,
```

Comment:
Please write it once, if you want, you can do explicit instantiation in the .cc and .cu files.

Reply:
The thing is, that there is supposed to be `float16` type in TYPE_SWITCH and it's impossible to write it that way. It's not there now, since we didn't decided what to do with it since we broke things up by introducing ifdeffed `float16`. But when we decide what to do, it'd be way easier to alternatively make it one function then the other way

Reply:
I don't see a logic of writing two copies of the same function because we may need to split it eventually, and if not, than we will unify it. Someone will forget about it or will wonder why those special cases for identical code.

Reply:
The same way is in `hsv.cc/cu`...